memory/iter.go

```diff
@@ -4,341 +4,250 @@ import (
-	bit := bt.Iter()
-	if reverse {
-		bit.Last()
-	} else {
-		bit.First()
-	}
-
-	return &baseIterator{
-		it:         bit,
-		upperBound: upperBound,
-		lowerBound: lowerBound,
-		reverse:    reverse,
+	if iter.it.Item().isDeleted {
+		iter.Next()
 	}
 }
 
-func (bit *baseIterator) Next() bool {
-	if bit.reverse {
-		return bit.it.Prev()
+func (iter *iterator) next() bool {
+	if iter.reverse {
+		// WARNING: There is a bug in `Prev` that can cause unexpected behaviour
+		// when attempting to iterate beyond the end of the iterator.
+		//
+		// This is documented by the test `TestBTreePrevBug`, and our current
+		// interface/implementation should prevent it from surfacing, but be careful
+		// with this call.
+		return iter.it.Prev()
 	}
-	return bit.it.Next()
+	return iter.it.Next()
 }
 
-func (bit *baseIterator) Prev() bool {
-	if bit.reverse {
-		return bit.it.Next()
-	}
-	return bit.it.Prev()
+func (iter *iterator) Key() []byte {
+	return iter.it.Item().key
 }
 
-func (bit *baseIterator) Item() dsItem {
-	return bit.it.Item()
+func (iter *iterator) Value() ([]byte, error) {
+	return iter.it.Item().val, nil
 }
 
-func (bit *baseIterator) Seek(key dsItem) bool {
-	if bit.reverse {
+func (iter *iterator) Seek(key []byte) {
+	// get the correct initial version for the seek
+	// if there exists an exact match in keys, use the latest version
+	// of that key, otherwise, use the provided DB version
+	// TODO this could use some "peek" mechanic instead of a full lookup
+	version := iter.version
+	result := iter.db.get(key, version)
+	if result.key != nil && !result.isDeleted {
+		version = result.version
+	}
+
+	if iter.reverse {
 		// Unfortunately the BTree iterator doesn't provide a reversed seek, so we have to
 		// do a bit of work ourselves here if iterating in reverse.
 
-		if bit.upperBound == nil {
+		if iter.end == nil {
 			// If no upper bound has been provided, e.g. via an `End` or `Prefix` option
 			// we can just return the last item in the BTree.
-			return bit.it.Last()
-		}
-
-		hasItems := bit.it.Seek(dsItem{key: bit.upperBound, version: key.version})
-		if hasItems {
-			// If the BTree iterator `Seek` finds an item, it must be equal or greater than
-			// our upper bound.  The previous item key must then be less than our upper bound.
-			hasItems = bit.it.Prev()
+			iter.hasItem = iter.it.Last()
+		} else {
+			iter.hasItem = iter.it.Seek(dsItem{key: iter.end, version: version})
+			if iter.hasItem {
+				// If the BTree iterator `Seek` finds an item, it must be equal or greater than
+				// our upper bound.  The previous item key must then be less than our upper bound.
+				iter.hasItem = iter.it.Prev()
+			}
+
+			if !iter.hasItem {
+				// If no items were found above or on the upper bound, we can move to the end of the
+				// BTree.
+				iter.hasItem = iter.it.Last()
+			}
+
+			if !iter.hasItem {
+				// If there are no items found at this point, it means the store is empty.
+				return
+			}
```

Comment:
todo: Reverse`Seek` is a bit broken. If I have `k1, k2, k3` and I want to seek to `k2` with or without specifying a specific end, it will return `k3` but I would expect `k2`.

You basically just need to add this at the top of the if-else statement to take into account the key match.
```go
if result.key != nil {
  iter.hasItem = iter.it.Seek(dsItem{key: result.key, version: version})
} else if iter.end == nil {
  ...
```

Reply:
todo: Also please add a unit test that shows it works as expected.

Reply:
Nice find! It definitely looks like it wont work correctly, which means something needs to change, thanks Fred :)

- [ ] Reverse Seek bug?

Reply:
There is still a small bug here. If i have `k1, k2, k3, k4` and I define the end as `k3` with `reverse = true`, The initial value will be `v2` but if I try to seek to `k4`, it will return `k3` with value `v3`. I would expect `k2` to be returned.

Reply:
`End` is inclusive, so returning `k3` is correct.

Reply:
Partially discussed over discord, continuing here.

`End` is only inclusive depending on how you look at `Reverse`.  Fred pointed out that `end` (implementation prop, not the option) should be exclusive if driven by a `Prefix` option, as otherwise it returns really weird results.

Badger behaviour is broken here and cannot be used for comparison https://github.com/sourcenetwork/corekv/issues/38.

I'm going to make `end`/`End` always exclusive for now, as skipping `End` is less weird than including `end` in a reversed `Prefix` iterator.  We can discuss the medium/long-term plan in a documentation-PR that I plan on opening soon (I plan on discussing a few more items there too).

- [ ] Make `end` exclusive